### PR TITLE
[SERVICE-820] Design app-facing directory API

### PR DESCRIPTION
**Pull request by bryangaleopenfin**
The best place to start with this PR is probably `directoryAdmin.ts`, the client API for modifying the distributed app directory. All other changes are either in direct support of this, or tidying things in close proximity. This includes:
- Modifying existing provider-side code to support namespaced app directories
- Adding a `RemoteDirectory` for integration testing the new API
- Adding additional mocks for unit testing

### CODE REVIEW

File: test/demo/utils/RemoteDirectory.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    public async add(arg: Parameters<T['add']>[0]): Promise<void> {
+        return ofBrowser.executeOnWindow(this._executionTarget, function (
+            this: TestWindowContext,
+            remoteIndex: number,
+            remoteAspect: T['type'],
+            remoteArg: Parameters<T['add']>[0]
+        ): void {
+            this.directories[remoteIndex][remoteAspect].add(remoteArg as any);
```
Comment (bryangaleopenfin) [inline test/demo/utils/RemoteDirectory.ts:61]: Open to ideas on how to get rid of these `any`s

Comment (tomrobertsOF) [reply test/demo/utils/RemoteDirectory.ts:61]: I think we're getting beyond the limits of Typescript here. May as well leave as is.

File: test/demo/utils/globalSetup.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@ import * as path from 'path';
 
 import * as mkdirp from 'mkdirp';
 import fetch from 'node-fetch';
-import * as puppeteer from 'puppeteer';
+import puppeteer from 'puppeteer';
```
Comment (tomrobertsOF) [inline test/demo/utils/globalSetup.ts:7]: Why this change?

**Review by tomrobertsOF — COMMENTED** [review]
Just the one tiny comment. Otherwise all good.

File: test/demo/utils/globalSetup.ts (hunk 2 above)
Comment (bryangaleopenfin) [reply test/demo/utils/globalSetup.ts:7]: This meant I could get rid of the `"esModuleInterop": false` line from `test/tsconfig.json`, which was a problem for the `deep-equal` import

Comment (tomrobertsOF) [reply test/demo/utils/globalSetup.ts:7]: Well that's weird, but ok i guess

**Review by tomrobertsOF — APPROVED** [review]

File: src/client/api/directoryAdmin.ts [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+ * handler callback provided by the application. This callback will be called with an [[AppDirectory]] object which
+ * allows the current (possibly empty) state of the app directory to the read and modified.
+ *
+ * Where app directories of different names provide applications of conflicting names, appIds, or UUIDs, the service
+ * will arbitrarily choose one of the conflicting applications to include and discard the rest.
+ *
+ * The app directory for a given domain will have an associated version number, which will default to 1 when an app
+ * directory is first created. The [[updateAppDirectory]] function allows this version number to be updated and
+ * queried, which may be useful to your organization if you have several apps that may be trying to update their
+ * domain's app directory in potentially differing and depricated ways. It is up to your organization how you wish to
```
Comment (ngunday) [inline src/client/api/directoryAdmin.ts]: typo: depricated -> deprecated

File: src/client/api/directoryAdmin.ts [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+     * If this is specified, and the current version of the app directory is greater than this version, the migration
+     * handler will not be called and the directory will not be modified.
+     */
+    maxSourceVersion?: number;
+
+    /**
+     * The version of the app directory outputted by the migration handler. If exists, must be a positive integer,
+     * and should be greater than or equal to `maxSourceVersion` if that also exists. If this is specified, and the
+     * current version of the app directory is greater than this version, the migration handler will not be called and
+     * the directory will not be modified. If this is not specied, the version of the app directory will be unchanged,
```
Comment (ngunday) [inline src/client/api/directoryAdmin.ts]: typo: specied -> specified

File: src/client/api/directoryAdmin.ts [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+export interface AppDirectory {
+    /**
+     * The current version of the app directory, or 0 if no app directory has yet been created.
+     */
+    sourceVersion: number;
+
+    /**
+     * A [[DirectoryCollection]] object that allows reading and modifying of remote snippets in the current app
+     * directory. Remote snippets are URLs that contain a JSON array of [[Application]]s to be included in the app
+     * directory. URLs not within the current appliction's domain will be ignored by the service.
```
Comment (ngunday) [inline src/client/api/directoryAdmin.ts]: typo: appliction -> application

File: src/client/api/directoryAdmin.ts [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+     * Adds either a single value or multiple values to this collection. Any duplicate values will be ignored.
+     *
+     * @param arg The value or array of values to add
+     */
+    add(arg: T | T[]): void;
+
+    /**
+     * Removes either a single value or multiple values from this collection. The value or values to be removed may be
+     * specified by either a value of type `T` that is deep-equal to the value to be removed, or a value of type `U`
+     * that refers to the value to be removed (implementation specific to the implentation of this interface).
```
Comment (ngunday) [inline src/client/api/directoryAdmin.ts]: typo: implentation -> implementation

File: src/client/api/directoryAdmin.ts [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+
+/**
+ * Interface for reading and modifying stored applications. When removing applications, applications can be refered to
+ * by either a complete, deep-equal [[Application]] object, or by the application's [[AppName]].
+ */
+export interface StoredApplicationsDirectoryCollection extends DirectoryCollection<Application, string> {
+    /**
+     * Adds the current application to the app directory.
+     *
+     * @param application A partial [[Application]] of any values that should override those derrived by the service
```
Comment (ngunday) [inline src/client/api/directoryAdmin.ts]: typo: derrived -> derived

File: test/client/api/directoryAdmin.unittest.ts [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+        await updateAppDirectory(spy, {
+            namespace: 'namespace-2',
+            maxSourceVersion: populatedDirectoryVersion - 1
+        });
+
+        expect(spy).not.toBeCalled();
+        expect(fin.Storage.setItem).not.toBeCalled();
+    });
+
+    test('If the specificed max version is equal to version of the app directory, the migration function is called \
```
Comment (ngunday) [inline test/client/api/directoryAdmin.unittest.ts]: typo: specificed -> specified

Comment (ngunday) [reply test/client/api/directoryAdmin.unittest.ts]: there are 3 more of this typo in this file.

File: test/client/api/directoryAdmin.unittest.ts [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+import {updateAppDirectory, AppDirectory, APP_DIRECTORY_STORAGE_TAG, StoredDirectoryShardMap, UpdateAppDirectoryMigrationHandler} from '../../../src/client/api/directoryAdmin';
+import {createMockFin, createMockApplication, getterMock} from '../../mocks';
+import {createFakeIdentity, createFakeApp, createFakeUrl} from '../../demo/utils/fakes';
+import {Application} from '../../../src/client/main';
+
+type TestParam = [string, UpdateAppDirectoryMigrationHandler, StoredDirectoryShardMap];
+
+const fin = createMockFin();
+
+const fakeApp1 = createFakeApp();
```
Comment (ngunday) [inline test/client/api/directoryAdmin.unittest.ts]: would it look less verbose if you kept these in an array?

File: src/client/api/directoryAdmin.ts [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+     * Removes all values from this collection, and then populates it with the value or values provided. Any duplicate
+     * values will be ignored.
+     *
+     * @param arg The value or array of values to populate this collection with
+     */
+    set(arg: T | T[]): void;
+}
+
+/**
+ * Inteface for reading and modifying remote snippets in the current app directory. Since these are specified as
```
Comment (ngunday) [inline src/client/api/directoryAdmin.ts]: typo: Inteface -> Interface

File: src/client/api/directoryAdmin.ts [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+ * `string`s there is no separate ID type.
+ */
+export interface RemoteSnippetsDirectoryCollection extends DirectoryCollection<string, string> {
+
+}
+
+/**
+ * Interface for reading and modifying stored applications. When removing applications, applications can be refered to
```
Comment (ngunday) [inline src/client/api/directoryAdmin.ts]: typo: refered -> referred

Comment (bryangaleopenfin) [reply src/client/api/directoryAdmin.ts]: ✅

File: src/client/api/directoryAdmin.ts (hunk 10 above)
Comment (bryangaleopenfin) [reply src/client/api/directoryAdmin.ts]: ✅

File: src/client/api/directoryAdmin.ts (hunk 3 above)
Comment (bryangaleopenfin) [reply src/client/api/directoryAdmin.ts]: ✅

File: src/client/api/directoryAdmin.ts (hunk 4 above)
Comment (bryangaleopenfin) [reply src/client/api/directoryAdmin.ts]: ✅

File: src/client/api/directoryAdmin.ts (hunk 6 above)
Comment (bryangaleopenfin) [reply src/client/api/directoryAdmin.ts]: ✅

File: src/client/api/directoryAdmin.ts (hunk 5 above)
Comment (bryangaleopenfin) [reply src/client/api/directoryAdmin.ts]: ✅

File: src/client/api/directoryAdmin.ts (hunk 7 above)
Comment (bryangaleopenfin) [reply src/client/api/directoryAdmin.ts]: ✅

File: test/client/api/directoryAdmin.unittest.ts (hunk 8 above)
Comment (bryangaleopenfin) [reply test/client/api/directoryAdmin.unittest.ts]: ✅

✅✅✅

File: test/client/api/directoryAdmin.unittest.ts (hunk 9 above)
Comment (bryangaleopenfin) [reply test/client/api/directoryAdmin.unittest.ts]: Probably not what you had in mind, but I've reduced the duplication here. Resistant to just keeping these in arrays for readability of the tests themselves

**Comment by bryangaleopenfin** [comment]
Back to you!

**Comment by bryangaleopenfin** [comment]
(Storage tag changed by request of runtime)

**Review by ngunday — APPROVED** [review]
